src/include/detail/ivf/qv.h

```diff
@@ -244,10 +163,185 @@ auto qv_query_heap_infinite_ram(
         });
   }
 
-  auto top_k = get_top_k_with_scores(min_scores, k_nn);
-  return top_k;
+  if constexpr (std::is_same_v<U, with_updates>) {
+    // @todo once working, refactor into separate function
+    auto num_adds = size(update_validity);
+    auto num_invalid = std::count(
+        update_validity.begin(),
+        update_validity.end(),
+        decltype(update_validity[0]){0});
+    auto num_valid = num_adds - num_invalid;
+    auto inserts = decltype(updates)(num_valid);
+    auto deletes = decltype(updates)(num_invalid);
+    auto insert_ids = decltype(update_ids)(num_valid);
+    auto delete_ids = decltype(update_ids)(num_invalid);
+
+    size_t insert_idx = 0;
+    size_t delete_idx = 0;
+    for (size_t i = 0; i < size(update_validity); ++i) {
+      if (update_validity[i]) {
+        inserts[insert_idx] = updates[i];
+        insert_ids[insert_idx] = update_ids[i];
+        ++insert_idx;
+      } else {
+        deletes[delete_idx] = updates[i];
+        delete_ids[delete_idx] = update_ids[i];
+        ++delete_idx;
+      }
+      assert (insert_idx <= num_valid);
+      assert (delete_idx <= num_invalid);
+      assert (insert_idx + delete_idx == i + 1);
+    }
+
+    filter_top_k_scores(min_scores, delete_ids);
```

Comment:
The IVF result should be filter by all update_ids and not just deletion_ids. This is because there can be changes in vector values for already existing ids and in these cases we only need to consider the vector once in the flat query bellow.